pkg/gator/sync/test/test.go

```diff
@@ -0,0 +1,148 @@
+package test
+
+import (
+	"fmt"
+
+	cfapis "github.com/open-policy-agent/frameworks/constraint/pkg/apis"
+	"github.com/open-policy-agent/frameworks/constraint/pkg/core/templates"
+	gkapis "github.com/open-policy-agent/gatekeeper/v3/apis"
+
+	gvkmanifestv1alpha1 "github.com/open-policy-agent/gatekeeper/v3/apis/gvkmanifest/v1alpha1"
+	"github.com/open-policy-agent/gatekeeper/v3/pkg/cachemanager/parser"
+	"github.com/open-policy-agent/gatekeeper/v3/pkg/gator/reader"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+)
+
+var scheme *runtime.Scheme
+
+func init() {
+	scheme = runtime.NewScheme()
+	err := cfapis.AddToScheme(scheme)
+	if err != nil {
+		panic(err)
+	}
+	err = gkapis.AddToScheme(scheme)
+	if err != nil {
+		panic(err)
+	}
+}
+
+// Reads a list of unstructured objects and a string containing supported GVKs and
+// outputs a set of missing sync requirements per template and ingestion problems per template.
+func Test(unstrucs []*unstructured.Unstructured, omitGVKManifest bool) (map[string]parser.SyncRequirements, map[string]error, error) {
+	templates := map[*templates.ConstraintTemplate]parser.SyncRequirements{}
+	syncedGVKs := map[schema.GroupVersionKind]struct{}{}
+	templateErrs := map[string]error{}
+	hasConfig := false
+	var gvkManifest *gvkmanifestv1alpha1.GVKManifest
+	var err error
+
+	for _, obj := range unstrucs {
+		if reader.IsSyncSet(obj) {
```

Comment:
You don't have to change it, but food for thought:

You could use the [GVKAggregator](https://github.com/open-policy-agent/gatekeeper/blob/master/pkg/cachemanager/aggregator/aggregator.go) type that Alex already built to help you gather the GVKs.  Since that is well tested to aggregate multiple SyncSets and a Config successfully, it could save you some testing below that checks a SyncSet vs. a Config.

Reply:
Do you mind elaborating on how that simplifies the step of ingesting an unstructured? I only saw it used in the context of SyncSet controller, where the controller was the one doing the syncset-specific watch and ingestion

Reply:
Ah, I read too far into the line of code that was commented on and assumed you were suggesting this would eliminate the IsSyncSet/ToSyncSet calls as well. Yeah I can go ahead and switch out my map with the gvkaggregator!